sdk/lib/opentelemetry/sdk/trace/samplers/consistent_probability_based.rb

```diff
@@ -0,0 +1,89 @@
+# frozen_string_literal: true
+
+# Copyright The OpenTelemetry Authors
+#
+# SPDX-License-Identifier: Apache-2.0
+
+module OpenTelemetry
+  module SDK
+    module Trace
+      module Samplers
+        # @api private
+        #
+        # Implements consistent sampling based on a probability.
+        class ConsistentProbabilityBased
+          include ConsistentProbabilityTraceState
+
+          attr_reader :description
+
+          def initialize(probability)
+            # TODO: probability should be 0 if it is < 2**-62
+            @probability = probability
+            @p_floor = (Math.frexp(probability)[1] - 1).abs
+            @p_ceil = @p_floor + 1
+            floor = Math.ldexp(1.0, -@p_floor)
+            ceil = Math.ldexp(1.0, -@p_ceil)
+            @p_ceil_probability = (probability - floor) / (ceil - floor)
+            @description = format('ConsistentProbabilityBased{%.6f}', probability)
+          end
+
+          def ==(other)
+            @description == other.description
+          end
+
+          # @api private
+          #
+          # See {Samplers}.
+          def should_sample?(trace_id:, parent_context:, links:, name:, kind:, attributes:) # rubocop:disable Metrics/MethodLength, Metrics/PerceivedComplexity
+            parent_span_context = OpenTelemetry::Trace.current_span(parent_context).context
+            if !parent_span_context.valid?
+              r = generate_r(trace_id)
+              p = probabilistic_p
+              if p < r
+                tracestate = TraceState.from_hash({ 'p' => p.to_s, 'r' => r.to_s })
+                Result.new(decision: Decision::RECORD_AND_SAMPLE, tracestate: tracestate)
+              else
+                tracestate = TraceState.from_hash({ 'r' => r.to_s })
+                Result.new(decision: Decision::DROP, tracestate: tracestate)
+              end
+            else
+              decision = nil
+              tracestate = validate_tracestate(parent_span_context) do |ot, parent_r|
+                if parent_r.nil?
+                  # TODO: warn the user that a potentially inconsistent trace is being produced
+                  r = generate_r(trace_id)
+                  ot.set_value('r', r.to_s)
+                end
+                p = probabilistic_p
+                if p < r
+                  ot.set_value('p', p.to_s)
+                  decision = Decision::RECORD_AND_SAMPLE
+                else
+                  decision = Decision::DROP
+                end
+                ot
+              end
+              Result.new(decision: decision, tracestate: tracestate)
+            end
+          end
+
+          private
+
+          def generate_r(trace_id)
+            x = trace_id[8, 8].unpack1('Q>') | 0x3
+            clz = 64 - x.bit_length
+            clz
+          end
```

Comment:
Yeah - I experimented with this a bunch in the interpreter. It's not intuitive, but it *is* efficient.

Reply:
To be clear, `clz` is "count leading zero bits", which is what the spec recommends.
```ruby
            x = trace_id[8, 8].unpack1('Q>') | 0x3
```
Take the same 8 bytes (64 bits) of the randomly generated trace ID that we use in the trace ID ratio-based sampler. Set the 2nd bit (well, 1st and 2nd). This ensures that when we count leading zeros, we get a max of 62 (64 - 2).
```ruby
            clz = 64 - x.bit_length
```
`bit_length` returns the minimum number of bits required to represent the value. I.e. a value like `0001 0011` has a bit length of 5 because we can represent it as `1 0011`. Since we have 64 bits and `bit_length` returns the highest set bit, `64 - x.bit_length` is the count of leading zeros.

Reply:
Ahhhhh this makes it much more clear. I really appreciate that, thank you. 😄